README.rst
InfluxGraph
=================

An `InfluxDB`_ storage plugin for `Graphite-API`_. Graphite with InfluxDB data store from any kind of schema(s) used in the DB.

.. image:: https://img.shields.io/pypi/v/influxgraph.svg
  :target: https://pypi.python.org/pypi/influxgraph
  :alt: Latest Version
.. image:: https://travis-ci.org/InfluxGraph/influxgraph.svg?branch=master
  :target: https://travis-ci.org/InfluxGraph/influxgraph
  :alt: CI status
.. image:: https://coveralls.io/repos/InfluxGraph/influxgraph/badge.png?branch=master
  :target: https://coveralls.io/r/InfluxGraph/influxgraph?branch=master
  :alt: Coverage
.. image:: https://readthedocs.org/projects/influxgraph/badge/?version=latest
  :target: http://influxgraph.readthedocs.io/en/latest/?badge=latest
  :alt: Documentation Status
.. image:: https://img.shields.io/pypi/wheel/influxgraph.svg
   :target: https://pypi.python.org/pypi/influxgraph
.. image:: https://img.shields.io/pypi/pyversions/influxgraph.svg
   :target: https://pypi.python.org/pypi/influxgraph


This project started as a re-write of `graphite influxdb <https://github.com/vimeo/graphite-influxdb>`_, now a separate project.


Installation
=============

Docker Compose
---------------

In `compose directory <https://github.com/InfluxGraph/influxgraph/tree/master/docker/compose>`_ can be found docker-compose configuration that will spawn all necessary services for a complete monitoring solution with:

* InfluxDB
* Telegraf
* Graphite API with InfluxGraph
* Grafana dashboard

To use, within compose directory run:

.. code-block:: shell

   docker-compose up

Grafana will be running on ``http://localhost:3000`` with Graphite datasource for InfluxDB data available at ``http://localhost``. Add a new Graphite datasource to Grafana - default Grafana user/pass is admin/admin - to create dashboards with.

See `compose configuration readme <https://github.com/InfluxGraph/influxgraph/tree/master/docker/compose>`_ for more details.

Docker Image
-------------

.. code-block:: shell

  docker pull ikuosu/influxgraph
  docker create  --name=influxgraph -p 8000:80 ikuosu/influxgraph
  docker start influxgraph

There will now be a Graphite-API running on ``localhost:8000`` from the container with a default InfluxDB configuration and memcache enabled. Finder expects InfluxDB to be running on ``localhost:8086`` by default.

The image will use a supplied ``graphite-api.yaml`` on build, when ``docker build`` is called on an InfluxGraph image.

`Docker file <https://github.com/InfluxGraph/influxgraph/blob/master/docker/Dockerfile>`_ used to build container can be found under ``docker`` directory of the repository.

.. note::

  If having issues with the container accessing the host's InfluxDB service then either use ``--network="host"`` when launching the container or build a new image with a provided configuration file containing the correct `InfluxDB host:port <https://github.com/InfluxGraph/influxgraph/blob/master/docker/graphite-api.yaml#L4>`_ destination.

Manual Installation
---------------------

.. code-block:: shell

  pip install influxgraph

Use of a local `memcached` service is highly recommended - see configuration section on how to enable.

Mimimal configuration for Graphite-API is below. See `Full Configuration Example`_ for all possible configuration options.

``/etc/graphite-api.yaml``

.. code-block:: yaml

  finders:
    - influxgraph.InfluxDBFinder

See the `Wiki <https://github.com/InfluxGraph/influxgraph/wiki>`_ and `Configuration`_ section for details.

.. contents:: Table of Contents

Main features
==============

* InfluxDB Graphite template support - expose InfluxDB tagged data as Graphite metrics with configurable metric paths
* Dynamically calculated group by intervals based on query date/time range - fast queries regardless of the date/time they span
* Configurable per-query aggregation functions by regular expression pattern
* Configurable per-query retention policies by query date/time range. Automatically use pre-calculated downsampled data in a retention policy for historical data
* Fast in-memory index for Graphite metric path queries as a Python native code extension
* Multi-fetch enabled - fetch data for multiple metrics with one query to InfluxDB
* Memcached integration
* Python 3 and PyPy compatibility
* Good performance even with extremely large number of metrics in the DB - generated queries are guaranteed to have ``O(1)`` performance characteristics

Google User's Group
=====================

There is a `Google user's group for discussion <https://groups.google.com/forum/#!forum/influxgraph>`_ which is open to the public.

Goals
======

* InfluxDB as a drop-in replacement data store to the Graphite query API
* Backwards compatibility with existing Graphite API clients like Grafana and Graphite installations migrated to InfluxDB data stores using Graphite input service *with or without* Graphite template configuration
* Expose native InfluxDB line protocol ingested data via the Graphite API
* Clean, readable code with complete documentation for public endpoints
* Complete code coverage with both unit and integration testing. Code has `>90%` test coverage and is integration tested against a real InfluxDB service
* Good performance at large scale. InfluxGraph is used in production with good performance on InfluxDB nodes with cardinality exceeding 5M and a write rate of over 5M metrics/minute or 66K/second.

The first three goals provide both

- A backwards compatible migration path for existing Graphite installations to use InfluxDB as a drop-in storage back-end replacement with no API client side changes required, meaning existing Grafana or other dashboards continue to work as-is.
- A way for native InfluxDB collection agents to expose their data via the *Graphite API* which allows the use of any Graphite API talking tool, the plethora of Graphite API functions, custom functions, functions across series, multi-series plotting and functions via Graphite glob expressions et al.

As of this time of writing, no alternatives exist with similar functionality, performance and compatibility.

Non-Goals
==========

* Graphite-Web support from the official Graphite project

Dependencies
=============

With the exception of `InfluxDB`_ itself, the other dependencies are installed automatically by ``pip``.

* ``influxdb`` Python module
* `Graphite-API`_
* ``python-memcached`` Python module
* `InfluxDB`_ service, versions ``1.0`` or higher

InfluxDB Graphite metric templates
==================================

`InfluxGraph` can make use of any InfluxDB data and expose them as Graphite API metrics, as well as make use of Graphite metrics added to InfluxDB as-is sans tags.

Even data written to InfluxDB by native InfluxDB API clients can be exposed as Graphite metrics, allowing transparent to clients use of the Graphite API with InfluxDB acting as its storage back-end.

To make use of tagged InfluxDB data, the finder needs to know how to generate a Graphite metric path from the tags used by InfluxDB.

The easiest way to do this is to use the Graphite service in InfluxDB with configured templates which can be used as-is in `InfluxGraph`_ configuration - see `Full Configuration Example`_ section for details. This presumes existing collection agents are using the Graphite line protocol to write to InfluxDB via its Graphite input service.

If, on the other hand, native `InfluxDB`_ metrics collection agents like `Telegraf <https://www.influxdata.com/time-series-platform/telegraf/>`_ are used, that data can too be exposed as Graphite metrics by writing appropriate template(s) in Graphite-API configuration alone.

See `Telegraf default configuration template <https://github.com/InfluxGraph/influxgraph/wiki/Telegraf-default-configuration-template>`_ for an example of this.

By default, the storage plugin makes no assumptions that data is tagged, per InfluxDB default Graphite service template configuration as below::
  
  [[graphite]]
    <..>
    # templates = []


Retention policy configuration
==============================

Pending implementation of a feature request that will allow InfluxDB to select and/or merge results from down-sampled data as appropriate, retention policy configuration is needed to support the use-case of down-sampled data being present in non default retention policies:

.. code-block:: yaml

  retention_policies:
      <time interval of query>: <retention policy name>

For example, to make a query with a group by interval of one minute or less, interval above one and less than thirty minutes and interval thirty minutes or above use the retention policies named ``default``, ``10min`` and ``30min`` respectively:

.. code-block:: yaml

  retention_policies:
      60: default
      600: 10min
      1800: 30min

While not required, retention policy interval is best kept close to or identical to ``deltas`` interval for best influx query performance.

See `Full Configuration Example`_ file for additional details.

Configuration
=======================

Minimal Configuration
----------------------

In graphite-api config file at ``/etc/graphite-api.yaml``:

.. code-block:: yaml

  finders:
    - influxgraph.InfluxDBFinder

The folowing default Graphite-API configuration is used if not provided:

.. code-block:: yaml

  influxdb:
     db: graphite

Full Configuration Example
---------------------------

See `Graphite-API example configuration file <https://github.com/InfluxGraph/influxgraph/blob/master/graphite-api.yaml.example>`_ for a complete configuration example.

Aggregation function configuration
-----------------------------------

The finder supports configurable aggregation and selector functions to use per metric path regular expression pattern. This is the equivalent of ``storage-aggregation.conf`` in Graphite's ``carbon-cache``.

Default aggregation function used is ``mean`` if no configuration provided nor any matching configuration.

InfluxGraph has pre-defined aggregation configuration matching ``carbon-cache`` defaults, namely:

.. code-block:: yaml

  aggregation_functions:
      \.min$ : min
      \.max$ : max
      \.last$ : last
      \.sum$ : sum

Defaults are overridden if ``aggregation_functions`` is configured in ``graphite-api.yaml`` as shown in configuration example.

An error will be printed to stderr if a configured aggregation function is not a known valid InfluxDB aggregation or selector method per `InfluxDB function list <https://docs.influxdata.com/influxdb/v1.1/query_language/functions/>`_.

Transformation functions, for example ``derivative``, may _not_ be used as they require a separate aggregation to be performed. Transformations are performed by Graphite-API instead, which also supports pluggable functions.

Known InfluxDB aggregation and selector functions are defined at ``influxgraph.constants.INFLUXDB_AGGREGATIONS`` and can be overriden if necessary.

.. note::

   When querying identical fields from multiple measurements InfluxDB allows only *one* aggregation function to be used for all identical fields in the query.
   
   In other words, make sure all identical InfluxDB fields matched by a Graphite query pattern, for example ``my_host.cpu.*.*`` have the same aggregation function configured.

   When using neither tagged data nor template configuration, the InfluxDB field to be queried is always ``value``. This is the case where this limitation is (most) relevant.

   ``InfluxGraph`` will use the first aggregation function configured and log a warning message to that effect if a pattern query resolves to multiple aggregation functions.


Memcached InfluxDB
------------------------

Memcached can be used to cache InfluxDB data so the `Graphite-API` can avoid querying the DB if it does not have to.

TTL configuration for memcache as shown in `Full Configuration Example`_ is only for InfluxDB series list with data query TTL set to the grouping interval used.

For example, for a query spanning twenty-four hours, a group by interval of one minute is used by default. TTL for memcache is set to one minute for that query.

For a query spanning one month, a fifteen minute group by interval is used by default. TTL is also set to fifteen minutes for that query.

Calculated intervals
--------------------

A data ``group by`` interval is automatically calculated depending on the date/time range of the query. This keeps data size in check regardless of query range and speeds up graph generation for large ranges.

Default configuration mirrors what `Grafana`_ uses with the native InfluxDB API.

Overriding the automatically calculated intervals can be done via the optional ``deltas`` configuration. See `Full Configuration Example`_ file for all supported configuration options.

Unlike other Graphite compatible data stores, InfluxDB performs aggregation on data query, not on ingestion. Queries made by InfluxGraph are therefore always aggregation queries with a group by clause.

Users that wish to retrieve all, non-aggregated, data points regardless of date/time range are advised to query `InfluxDB`_ directly.

Varnish caching InfluxDB API
----------------------------

The following is a sample configuration of `Varnish`_ as an HTTP cache in front of InfluxDB's HTTP API. It uses Varnish's default TTL of 60 sec for all InfluxDB queries.

The intention is for a local (to InfluxDB) Varnish service to cache frequently accessed data and protect the database from multiple identical requests, for example multiple users viewing the same dashboard.

InfluxGraph configuration should use Varnish port to connect to InfluxDB.

Unfortunately, given that clients like Grafana use POST requests for querying the Graphite API, which cannot be cached, using Varnish in front of a Graphite-API webapp would have no effect. Multiple requests for the same dashboard/graph will therefore still hit Graphite-API, but with Varnish in front of InfluxDB the more sensitive DB is spared from duplicated queries.

Substitute the default ``8086`` backend port with the InfluxDB API port for your installation if needed:

.. code-block:: tcl

  backend default {
    .host = "127.0.0.1";
    .port = "8086";
  }

  sub vcl_recv {
    unset req.http.cookie;
  }

Graphite API example configuration:

.. code-block:: yaml

  finders:
    - influxgraph.InfluxDBFinder
  influxdb:
    port: <varnish port>

Where ``<varnish_port>`` is Varnish's listening port.

Any other HTTP caching service will similarly work just as well.

Optional C Extensions
======================

In order of fastest to slowest, here is how the supported interpreters fare with and without C extensions. How much faster depends largely on hardware and compiler used - can expect at least `15x` and `4x` performance increases for CPython with extensions and PyPy respectively compared to standard CPython without extensions.

CPython with extensions will also use about `20x` less memory for the index than either PyPy or CPython without extensions.

#. CPython with C extensions
#. Pypy
#. CPython

There are two performance tests in the repository that can be used to see relative performance with and without extensions, for `index <https://github.com/InfluxGraph/influxgraph/blob/master/tests/index_perf.py>`_ and `template <https://github.com/InfluxGraph/influxgraph/blob/master/tests/templates_parse_perf.py>`_ functionality respectively. On PyPy extensions are purposefully disabled.

Known Limitations
===================

Data *fill* parameter and counter values
-----------------------------------------

*Changed in version 1.3.6*

As of version ``1.3.6``, the default *fill* parameter is **null** so as to not add values that do not exist in data - was ``previous`` in prior versions.

This default will break derivative calculated counter values when data sampling rate exceeds configured interval for the query - see `Calculated intervals`_.

For example, with a data sampling rate of sixty (60) seconds and default ``deltas`` configuration, queries of thirty (30) minutes and below will use a thirty (30) second interval and will contain null datapoints. This in turn causes Graphite functions like ``derivative`` and ``non_negative_derivative`` to only contain null datapoints.

The fill parameter is configurable - see `Full Configuration Example`_ - but is currently common for all metric paths.

For ``derivative`` and related functions to work, either set ``deltas`` configuration to not go below data sampling rate or set *fill* configuration to ``previous``.

Index for Graphite metric paths
--------------------------------

The index implementation via native code extension releases Python's GIL as much as possible, however, there will still be a response time increase while index is being re-built.

Without extensions response time increase will be much higher - building with extensions is highly recommended.

That said, building extensions can be disabled by running `setup.py` with the `DISABLE_INFLUXGRAPH_CEXT=1` environment variable set. A notice will be displayed by `setup.py` that extensions have been disabled.

Note that without native extension, performance is much lower and memory use of index much higher.

.. _Varnish: https://www.varnish-cache.org/
.. _Graphite-API: https://github.com/brutasse/graphite-api
.. _Grafana: https://github.com/grafana/grafana
.. _InfluxDB: https://github.com/influxdb/influxdb


Changelog.rst
Change Log
=============

1.5.0
++++++

Changes
--------

* Series parsing C implementation for `2x` - `10x` index build speedup
* Configurable memcache series loader and fields keys


1.4.0
++++++

Release includes binary wheels for Linux and OSX.

Changes
--------

* Config name of DB password - #39
* Add config option for index lock file path - #37


graphite_influxdb.py
# Copyright (C) [2014-2015] [Vimeo, LLC]
# Copyright (C) [2015-] [Panos Kittenis]

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Graphite-Api storage finder for InfluxDB.

Read metric series from an InfluxDB database via a Graphite compatible API."""

import datetime
from influxdb import InfluxDBClient
import logging
from logging.handlers import TimedRotatingFileHandler
import re
import sys
from graphite_api.node import LeafNode, BranchNode
try:
    import statsd
except ImportError:
    pass

logger = logging.getLogger('graphite_influxdb')
# logging.basicConfig()
# logger.setLevel(logging.DEBUG)

# Tell influxdb to return time as seconds from epoch
_INFLUXDB_CLIENT_PARAMS = {'epoch': 's'}


class NullStatsd(object):
    """Fake StatsClient compatible class to use when statsd is not configured"""

    def __enter__(self):
        return self

    def __exit__(self, _type, value, traceback):
        pass

    def timer(self, key, val=None):
        return self

    def timing(self, key, val):
        pass

    def start(self):
        pass

    def stop(self):
        pass


def normalize_config(config):
    cfg = config.get('influxdb', None)
    ret = {}
    if not cfg:
        logger.critical("Missing required 'influxdb' configuration in graphite-api"
                        "config - please update your configuration file to include"
                        "at least 'influxdb: db: <db_name>'")
        sys.exit(1)
    ret['host'] = cfg.get('host', 'localhost')
    ret['port'] = cfg.get('port', 8086)
    ret['user'] = cfg.get('user', 'graphite')
    ret['passw'] = cfg.get('pass', 'graphite')
    ret['db'] = cfg.get('db', 'graphite')
    ssl = cfg.get('ssl', False)
    ret['ssl'] = (ssl == 'true')
    ret['schema'] = cfg.get('schema', [])
    ret['log_file'] = cfg.get('log_file', None)
    ret['log_level'] = cfg.get('log_level', 'info')
    cfg = config.get('es', {})
    if config.get('statsd', None):
        ret['statsd'] = config.get('statsd')
    return ret


def _make_graphite_api_points_list(influxdb_data):
    """Make graphite-api data points dictionary from Influxdb ResultSet data"""
    _data = {}
    for key in influxdb_data.keys():
        _data[key[0]] = [(datetime.datetime.fromtimestamp(float(d['time'])),
                          d['value']) for d in influxdb_data.get_points(key[0])]
    return _data


class InfluxdbReader(object):
    """Graphite-Api reader class for InfluxDB.
    
    Retrieves a single metric series from InfluxDB
    """
    __slots__ = ('client', 'path', 'step', 'statsd_client')

    def __init__(self, client, path, step, statsd_client):
        self.client = client
        self.path = path
        self.step = step
        self.statsd_client = statsd_client

    def fetch(self, start_time, end_time):
        """Fetch single series' data from > start_time and <= end_time
        
        :param start_time: start_time in seconds from epoch
        :param end_time: end_time in seconds from epoch
        """
        logger.debug("fetch() path=%s start_time=%s, end_time=%s, step=%d", self.path, start_time, end_time, self.step)
        timer_name = ".".join(['service_is_graphite-api',
                               'ext_service_is_influxdb',
                               'target_type_is_gauge',
                               'unit_is_ms',
                               'what_is_query_individual_duration'])
        with self.statsd_client.timer(timer_name):
            _query = 'select mean(value) as value from "%s" where (time > %ds and time <= %ds) GROUP BY time(%ss)' % (
                self.path, start_time, end_time, self.step)
            logger.debug("fetch() path=%s querying influxdb query: '%s'", self.path, _query)
            data = self.client.query(_query, params=_INFLUXDB_CLIENT_PARAMS)
        logger.debug("fetch() path=%s returned data: %s", self.path, data)
        try:
            data = _make_graphite_api_points_list(data)
        except Exception:
            logger.debug("fetch() path=%s COULDN'T READ POINTS. SETTING TO EMPTY LIST", self.path)
            data = []
        time_info = start_time, end_time, self.step
        return time_info, [v[1] for v in data[self.path]]
    
    def get_intervals(self):
        """Noop function - Used by Graphite-Web but not needed for Graphite-Api"""
        pass


class InfluxLeafNode(LeafNode):
    """Tell Graphite-Api that our leaf node supports multi-fetch"""
    __fetch_multi__ = 'influxdb'


class InfluxdbFinder(object):
    """Graphite-Api finder for InfluxDB.
    
    Finds and fetches metric series from InfluxDB.
    """
    __fetch_multi__ = 'influxdb'
    __slots__ = ('client', 'schemas', 'config', 'statsd_client')

    def __init__(self, config=None):
        config = normalize_config(config)
        self.config = config
        self.client = InfluxDBClient(config.get('host', 'localhost'),
                                     config.get('port', '8086'),
                                     config.get('user', 'root'),
                                     config.get('passw', 'root'),
                                     config['db'],
                                     config.get('ssl', 'false'),)
        self.schemas = [(re.compile(patt), step) for (patt, step) in config['schema']]
        try:
            self.statsd_client = statsd.StatsClient(config['statsd'].get('host'),
                                                    config['statsd'].get('port', 8125)) \
                if 'statsd' in config and config['statsd'].get('host') else NullStatsd()
        except NameError:
            logger.warning("Statsd client configuration present but 'statsd' module"
                           "not installed - ignoring statsd configuration..")
            self.statsd_client = NullStatsd()
        self._setup_logger(config['log_level'], config['log_file'])

    def _setup_logger(self, level, log_file):
        """Setup log level and log file if set"""
        if logger.handlers:
            return
        level = getattr(logging, level.upper())
        logger.setLevel(level)
        formatter = logging.Formatter(
            '[%(levelname)s] %(asctime)s - %(module)s.%(funcName)s() - %(message)s')
        handler = logging.StreamHandler()
        logger.addHandler(handler)
        handler.setFormatter(formatter)
        if not log_file:
            return
        try:
            handler = TimedRotatingFileHandler(log_file)
        except IOError:
            logger.error("Could not write to %s, falling back to stdout",
                         log_file)
        else:
            logger.addHandler(handler)
            handler.setFormatter(formatter)

    def get_series(self, query):
        """Retrieve series names from InfluxDB according to query pattern
        
        :param query: Query to run to get series names
        :type query: :mod:`graphite_api.storage.FindQuery` compatible class
        """
        # regexes in influxdb are not assumed to be anchored, so anchor them explicitly
        regex = self.compile_regex('^{0}', query)
        timer_name = ".".join(['service_is_graphite-api',
                               'ext_service_is_influxdb',
                               'target_type_is_gauge',
                               'unit_is_ms',
                               'action_is_get_series'])
        with self.statsd_client.timer(timer_name):
            _query = "show series from /%s/" % regex.pattern
            logger.debug("get_series() Calling influxdb with query - %s", _query)
            data = self.client.query(_query, params=_INFLUXDB_CLIENT_PARAMS)
            # as long as influxdb doesn't have good safeguards against
            # series with bad data in the metric names, we must filter out
            # like so:
            series = [key_name for (key_name, _) in data.keys()]
        return series

    def compile_regex(self, fmt, query):
        r"""Turn glob (graphite) queries into compiled regex.
        
        * becomes .*
        . becomes \.
        fmt argument is so that caller can control anchoring (must contain exactly one {0} !
        """
        return re.compile(fmt.format(
            query.pattern.replace('.', r'\.').replace('*', r'[^\.]*').replace(
                '{', '(').replace(',', '|').replace('}', ')')
        ))

    def get_leaves(self, query):
        """Get LeafNode according to query pattern

        :param query: Query to run to get LeafNodes
        :type query: :mod:`graphite_api.storage.FindQuery` compatible class
        """
        key_leaves = "%s_leaves" % query.pattern
        series = self.get_series(query)
        regex = self.compile_regex('^{0}$', query)
        logger.debug("get_leaves() key %s", key_leaves)
        timer_name = ".".join(['service_is_graphite-api',
                               'action_is_find_leaves',
                               'target_type_is_gauge',
                               'unit_is_ms'])
        timer = self.statsd_client.timer(timer_name)
        now = datetime.datetime.now()
        timer.start()
        # return every matching series and its
        # resolution (based on first pattern match in schema, fallback to 60s)
        leaves = [(name, next((res for (patt, res) in self.schemas if patt.match(name)), 60))
                  for name in series if regex.match(name)
                  ]
        timer.stop()
        end = datetime.datetime.now()
        dt = end - now
        logger.debug("get_leaves() key %s Finished find_leaves in %s.%ss",
                     key_leaves,
                     dt.seconds,
                     dt.microseconds)
        return leaves

    def get_branches(self, query):
        """Get branches according to query.

        :param query: Query to run to get BranchNodes
        :type query: :mod:`graphite_api.storage.FindQuery` compatible class
        """
        seen_branches = set()
        key_branches = "%s_branches" % query.pattern
        # Very inefficient call to list
        series = self.get_series(query)
        regex = self.compile_regex('^{0}$', query)
        logger.debug("get_branches() %s", key_branches)
        timer_name = ".".join(['service_is_graphite-api',
                               'action_is_find_branches',
                               'target_type_is_gauge',
                               'unit_is_ms'])
        timer = self.statsd_client.timer(timer_name)
        start_time = datetime.datetime.now()
        timer.start()
        branches = []
        for name in series:
            while '.' in name:
                name = name.rsplit('.', 1)[0]
                if name not in seen_branches:
                    seen_branches.add(name)
                    if regex.match(name) is not None:
                        logger.debug("get_branches() %s found branch name: %s", key_branches, name)
                        branches.append(name)
        timer.stop()
        end_time = datetime.datetime.now()
        dt = end_time - start_time
        logger.debug("get_branches() %s Finished find_branches in %s.%ss",
                     key_branches,
                     dt.seconds, dt.microseconds)
        return branches

    def find_nodes(self, query):
        """Find matching nodes according to query.

        :param query: Query to run to find either BranchNode(s) or LeafNode(s)
        :type query: :mod:`graphite_api.storage.FindQuery` compatible class
        """
        logger.debug("find_nodes() query %s", query)
        # TODO(retention_periods): once we can query influx better for retention periods,
        # honour the start/end time in the FindQuery object
        timer_name = ".".join(['service_is_graphite-api',
                               'action_is_yield_nodes',
                               'target_type_is_gauge',
                               'unit_is_ms.what_is_query_duration'])
        with self.statsd_client.timer(timer_name):
            for (name, res) in self.get_leaves(query):
                yield InfluxLeafNode(name, InfluxdbReader(
                    self.client, name, res, self.statsd_client))
            for name in self.get_branches(query):
                logger.debug("Yielding branch %s", name,)
                yield BranchNode(name)

    def fetch_multi(self, nodes, start_time, end_time):
        """Fetch datapoints for all series between start and end times

        :param nodes: List of nodes to retrieve data for
        :type nodes: list(:mod:`graphite_influxdb.InfluxdbLeafNode`)
        :param start_time: Start time of query
        :param end_time: End time of query
        """
        series = ', '.join(['"%s"' % node.path for node in nodes])
        # use the step of the node that is the most coarse
        # not sure if there's a better way? can we combine series
        # with different steps (and use the optimal step for each?)
        # probably not
        step = max([node.reader.step for node in nodes])
        query = 'select mean(value) as value from %s where (time > %ds and time <= %ds) GROUP BY time(%ss)' % (
                series, start_time, end_time, step)
        logger.debug('fetch_multi() query: %s', query)
        logger.debug('fetch_multi() - start_time: %s - end_time: %s, step %s',
                     datetime.datetime.fromtimestamp(float(start_time)),
                     datetime.datetime.fromtimestamp(float(end_time)), step)
        timer_name = ".".join(['service_is_graphite-api',
                               'ext_service_is_influxdb',
                               'target_type_is_gauge',
                               'unit_is_ms',
                               'action_is_select_datapoints'])
        with self.statsd_client.timer(timer_name):
            logger.debug("Calling influxdb multi fetch with query - %s", query)
            data = self.client.query(query, params=_INFLUXDB_CLIENT_PARAMS)
        logger.debug('fetch_multi() - Retrieved %d result set(s)', len(data))
        data = _make_graphite_api_points_list(data)
        # some series we requested might not be in the resultset.
        # this is because influx doesn't include series that had no values
        # this is a behavior that some people actually appreciate when graphing,
        # but graphite doesn't do this (yet),
        # and we want to look the same, so we must add those back in.
        # a better reason though, is because for advanced alerting cases like bosun,
        # you want all entries even if they have no data, so you can properly
        # compare, join, or do logic with the targets returned for requests for the
        # same data but from different time ranges, you want them to all
        # include the same keys.
        query_keys = set([node.path for node in nodes])
        for key in query_keys:
            data.setdefault(key, [])
        time_info = start_time, end_time, step
        for key in data:
            data[key] = [v[1] for v in data[key]]
        return time_info, data
